plugins/libplugin-pay.c

```diff
@@ -178,7 +178,38 @@ static u8 *tal_towire_legacy_payload(const tal_t *ctx, const struct legacy_paylo
 	return buf;
 }
 
-static struct command_result *payment_createonion_success(struct command *cmd,
+static struct createonion_response *
+tal_createonion_response_from_json(const tal_t *ctx, const char *buffer,
+				   const jsmntok_t *toks)
+{
+	size_t i;
+	struct createonion_response *resp;
+	const jsmntok_t *oniontok = json_get_member(buffer, toks, "onion");
+	const jsmntok_t *secretstok = json_get_member(buffer, toks, "shared_secrets");
+	const jsmntok_t *cursectok;
+
+	if (oniontok == NULL || secretstok == NULL)
+		return NULL;
+	resp = tal(ctx, struct createonion_response);
+
+	if (oniontok->type != JSMN_STRING)
+		goto fail;
+
+	resp->onion = json_tok_bin_from_hex(resp, buffer, oniontok);
+	resp->shared_secrets = tal_arr(resp, struct secret, secretstok->size);
+
+	json_for_each_arr(i, cursectok, secretstok) {
+		if (cursectok->type != JSMN_STRING)
+			goto fail;
+		json_to_secret(buffer, cursectok, &resp->shared_secrets[i]);
+	}
+	return resp;
+
+fail:
+	return tal_free(resp);
```

Comment:
... error messages on these failure paths?

Reply:
Handled in the caller. Since we're parsing an internal RPC method I think we can leave the error message generic.